// grammy/out/convenience/webhook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webhookCallback = void 0;
const platform_node_js_1 = require("../platform.node.js");
const debugErr = (0, platform_node_js_1.debug)("grammy:error");
const standard = (req, res) => ({
    update: Promise.resolve(req.body),
    end: () => res.end(),
    respond: (json) => {
        res.set("Content-Type", "application/json");
        res.send(json);
    },
});
// Integrations with popular frameworks
const frameworkAdapters = {
    express: standard,
    http: standard,
    https: standard,
    koa: (ctx) => ({
        update: Promise.resolve(ctx.request.body),
        end: () => (ctx.body = ""),
        respond: (json) => {
            ctx.set("Content-Type", "application/json");
            ctx.response.body = json;
        },
    }),
    oak: (ctx) => ({
        update: ctx.request.body({ type: "json" }).value,
        end: () => (ctx.response.status = 200),
        respond: (json) => {
            ctx.response.type = "json";
            ctx.response.body = json;
        },
    }),
    fastify: (req, reply) => ({
        update: Promise.resolve(req.body),
        end: () => reply.send({}),
        respond: (json) => reply.send(json),
    }),
    worktop: (req, res) => ({
        update: Promise.resolve(req.body.json()),
        end: () => res.end(),
        respond: (json) => res.send(200, json),
    }),
    callback: (update, callback) => ({
        update: update,
        respond: callback,
    }),
    "aws-lambda": (event, _context, callback) => ({
        update: JSON.parse(event.body),
        end: () => callback(null, { statusCode: 200 }),
        respond: (json) => callback(null, {
            statusCode: 200,
            body: json,
        }),
    }),
    // please open a PR if you want to add another
};
/**
 * Creates a callback function that you can pass to a web framework (such as
 * express) if you want to run your bot via webhooks. Use it like this:
 * ```ts
 * const app = express() // or whatever you're using
 * const bot = new Bot('<token>')
 *
 * app.use(webhookCallback(bot, 'express'))
 * ```
 *
 * Confer the grammY
 * [documentation](https://grammy.dev/guide/deployment-types.html) to read more
 * about how to run your bot with webhooks.
 *
 * @param bot The bot for which to create a callback
 * @param framework An optional string identifying the framework (default: 'express')
 * @param onTimeout An optional strategy to handle timeouts (default: 'throw')
 * @param timeoutMilliseconds An optional number of timeout milliseconds (default: 10_000)
 */
function webhookCallback(bot, framework = "express", onTimeout = "throw", timeoutMilliseconds = 10000) {
    var _a;
    const server = (_a = frameworkAdapters[framework]) !== null && _a !== void 0 ? _a : standard;
    let firstUpdate = true;
    let initialized = false;
    let initCall;
    return async (...args) => {
        const { update, respond, end } = server(...args);
        let usedWebhookReply = false;
        const webhookReplyEnvelope = {
            send: async (json) => {
                usedWebhookReply = true;
                await respond(json);
            },
        };
        if (!initialized) {
            if (firstUpdate) {
                initCall = bot.init();
                firstUpdate = false;
            }
            await initCall;
            initialized = true;
        }
        await timeoutIfNecessary(bot.handleUpdate(await update, webhookReplyEnvelope), typeof onTimeout === "function"
            ? () => onTimeout(...args)
            : onTimeout, timeoutMilliseconds);
        if (end !== undefined && !usedWebhookReply)
            end();
    };
}
exports.webhookCallback = webhookCallback;
function timeoutIfNecessary(task, onTimeout, timeout) {
    if (timeout === Infinity)
        return task;
    return new Promise((resolve, reject) => {
        const handle = setTimeout(() => {
            if (onTimeout === "throw") {
                reject(new Error(`Request timed out after ${timeout} ms`));
            }
            else {
                if (typeof onTimeout === "function")
                    onTimeout();
                resolve();
            }
            const now = Date.now();
            task.finally(() => {
                const diff = Date.now() - now;
                debugErr(`Request completed ${diff} ms after timeout!`);
            });
        }, timeout);
        task.then(resolve)
            .catch(reject)
            .finally(() => clearTimeout(handle));
    });
}
